Graph

danielarn opened the pull request.

danielarn committed:
Update Board.cs

BoardGraph basic structure

danielarn committed:
Basic Graph structure

cwrothrock commented on Assets/Scripts/Board/Board.cs:
```diff
@@ -79,6 +79,42 @@ public TerrainTile(Vector3Int position, TileType tileType, int diceNumber = 0)
+        public TileNode() { }
+    }
+
+    private class BoardGraph
+    {
+        private VertexNode[] vertexNodes;
+        private EdgeNode[] edgeNodes;
+        private TileNode[] tileNodes;
+
+        private BoardGraph() { }
+
+
+    }
```
Do you think we should just store these lists as members of the `Board` class rather than in this separate subclass? Not opposed to either initially
also nit: whitespace

danielarn replied:
That feels fine to me. Fixed.

danielarn committed:
Update Board.cs

danielarn committed:
Update Board.cs

cwrothrock commented on Assets/Scripts/Board/Board.cs:
```diff
@@ -79,6 +79,35 @@ public TerrainTile(Vector3Int position, TileType tileType, int diceNumber = 0)
+    private EdgeNode[] edgeNodes;
+    private TileNode[] tileNodes;
+    // end Graph
+
```
Lets make this a separate `Graph.cs` class which is not a MonoBehaviour and simply handle storing the structure and node ownership/connections. Board.cs can be the MonoBehaviour that unity uses to orchestrate gameplay and create a Graph object on start()

danielarn replied:
Done. Will add logic to use Graph to Board.cs later.

danielarn committed:
Move graph code to Graph.cs

cwrothrock commented on Assets/Scripts/Board/Graph.cs:
```diff
@@ -0,0 +1,80 @@
+
+public class Graph
+{
+    public enum Building
+    {
+        Settlement,
+        City,
+    }
```
We should add `NONE` so that doing `public Building building;` defaults to NONE, since I think enum-typed objects always default to 0th indexed value. But check me on that before making change, as long as we can distinguish between `building` being null/none and actually being set/occupied.

danielarn replied:
I'm fine with that. I had that originally, but removed for no real reason.

cwrothrock commented on Assets/Scripts/Board/Graph.cs:
```diff
@@ -0,0 +1,80 @@
+        City,
+    }
+
+    public class Node { }
```
The main advantage of having a Node base class is to either (1) place common logic between all nodes, or (2) make it possible to have a List<Node> that contains both VertexNode/EdgeNode objects. Maybe we don't have a reason to do these things yet and will in the future, but worth noting. 

(1) could include
* `Building` if `Building` included Road and Robber
* the Vector3Int and/or Vector3 which store the node's position, that way it can be accessed and used by the Board class to get the world position in the tilemap 

Just food for thought, nothing actionable per se and not sure if it's even a good thought.

danielarn replied:
I added a `worldPosition` field to `Node`, which feels helpful. As far as augmenting `Building` with road and robber, I'm not sure I like that, since the type of building that a node can have would then be limited by the node type, but that isn't represented by having all them in the `Building` enum.

cwrothrock commented on Assets/Scripts/Board/Graph.cs:
```diff
@@ -0,0 +1,80 @@
+    public class EdgeNode : Node
+    {
+        public List<VertexNode> vertices = new();
+        public bool hasRoad = false;
```
Using a bool switch for a one-time setter isn't great IMO. It makes sense for the TileNode somewhat in that it changes frequently, but roads will only ever change once, i.e. once the EdgeNode has an owner, that owner will never change, so we can just use the owner to determine if the edge is occupied by a road

danielarn replied:
Good point. Updated.

cwrothrock commented on Assets/Scripts/Board/Graph.cs:
```diff
@@ -0,0 +1,80 @@
+    public List<VertexNode> vertices = new();
+    public List<EdgeNode> edges = new();
+    public Dictionary<Vector3Int, TileNode> tiles = new();
+
```
Maybe these should be private? And then we only expose functionality via the public functions below

danielarn replied:
SGTM.

danielarn committed:
Update Graph.cs

Add `None` building type.
Make member variables private for classes.
Add more getters and setters.
